improver/psychrometric_calculations/hail_size.py

```diff
@@ -0,0 +1,392 @@
+
+    def variable_at_pressure(
+        self, variable_on_pressure: Cube, pressure: np.ndarray
+    ) -> np.ndarray:
+        """Extracts the values from variable_on_pressure cube at pressure
+        levels described by the pressure array"""
+
+        variable = []
+        pressure_grid = self.pressure_grid(variable_on_pressure)
+
+        # deals with each realization independently and puts the arrays back together
+        for rea in range(len(pressure_grid)):
+            pressure_diff = abs(pressure_grid[rea] - pressure[rea])
+
+            indicies = np.nanargmin(pressure_diff, axis=0)
+            lat, long = indicies.shape
+            lat, long = np.ogrid[:lat, :long]
+
+            variable.append(variable_on_pressure.data[rea, indicies, lat, long])
+
+        return np.array(variable)
+
+    def pressure_grid(self, variable_on_pressure: Cube) -> np.ndarray:
+        """Creates a pressure_grid of the same shape as variable_on_pressure
+        with all possible pressure levels."""
+
+        required_shape = variable_on_pressure.shape
+        pressure_points = variable_on_pressure.coord("pressure").points
+        (pressure_axis,) = variable_on_pressure.coord_dims("pressure")
+        pressure_shape = np.ones_like(required_shape)
+        pressure_shape[pressure_axis] = required_shape[pressure_axis]
+        pressure_array = np.broadcast_to(
+            pressure_points.reshape(pressure_shape), required_shape
+        )
+        return pressure_array
+
+    def pressure_at_268(self, temperature_on_pressure: Cube) -> (CubeList):
+        """Extracts the lowest pressure level where the environment
+        temperature is below -5 degrees (268.15K). It also produces
+        the environemnt temperature at that pressure value"""
+
+        pressure_template = next(temperature_on_pressure.slices_over(["pressure"]))
+        pressure_template.rename("pressure of atmosphere at 268.15K")
+        pressure_template.units = temperature_on_pressure.coord("pressure").units
+        pressure_template.remove_coord("pressure")
+
+        temperature_template = next(temperature_on_pressure.slices_over(["pressure"]))
+        temperature_template.rename("tempeature_of_atmosphere_at_268.15K")
+        temperature_template.remove_coord("pressure")
+
+        shape = temperature_on_pressure.data.shape
+        r, l, m = (shape[0], shape[2], shape[3])  # ignores pressure axis
+
+        data = np.ma.masked_greater(temperature_on_pressure.data, 268.15)
+        data = np.ma.masked_invalid(data)
+
+        indicies = np.ma.notmasked_edges(data, axis=1)[0][1]
+
+        pressure = temperature_on_pressure.coord("pressure").points[indicies]
+
+        # identifies if there are columns where the temperature never drops below -5 degrees
+        if len(pressure) != r * l * m:
+            index = np.where(np.amin(data, axis=1).flatten() > 268.15)[0]
+            for x in index:
+                pressure = np.insert(pressure, x, -9999)
```

Comment:
Does this work? Doesn't the flattened data array potentially have indices up to enormous numbers and the pressure coordinate is only (in our case) 33 elements long?

Reply:
I have actually updated this bit of code to try and make it a bit better and clearer but I think it does work as I intend.

This bit of code is needed in cases where there is atleast 1 grid square where the entire column has ended up masked because the temperatue never goes below -5 degrees or has invalid values. In this case the np.ma.notmasked_edges returns nothing. The list called pressure should have 1 pressure value for every grid square but if there was a masked grid square then the pressure array ends up being too short as its missing values.

This now goes through and finds the indexes where there are columns of entirely masked values and inserts -9999 into the array to make it the correct shape and then masks these points. It's not actually calling the pressure_coordinate so the fact the pressure_coordinate is 33 elements doesn't matter as the indicies are merely to know where to insert values into the array so they're associated with the correct grid squares. 

Hopefully that makes sense?